### PR TITLE
Always set `Sys.catch_break true`, not just when interactive

### PR DESCRIPTION
**Pull request by SkySkimmer**
When debugging it can be nice to run coqc and try to get a backtrace from interrupting it.

eg https://coq.zulipchat.com/#narrow/stream/237656-Coq-devs-.26-plugin-devs/topic/Debugging.20hangs

OTOH if we're hanging in such a way that we can't process signals this
may be a bad idea so maybe we should only enable it with backtraces in
non interactive mode?

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by ejgallego** [comment]
I'll look into this ASAP, but indeed this stuff is delicate, mainly due to us virtually having no tests for interrupting coq.


Note that still many users rely on proof general interrupting Coq properly, I am not sure this patch enables interrupting in undesired places.

(Flèche / cow-lsp were designed from scratch to handle interruptions in a principled way)

My preliminary though is that if this is useful in debugging, then it should take effect only in that case.

**Comment by ejgallego** [comment]
I did discuss with @gasche (thanks Gabriel!) about whether it would make sense to run `coqc` with `Sys.catch_break true` or `false`, so in the end, it shouldn't matter as the only difference is the signal handler that is installed.

Not even the aligment of code should change? [Tho coqinit change could impact in this sense]

The bench results are pretty surprising, let's see if they are reproducible.

@coqbot bench

[Is there a way to tell coqbot to re-bench only a set of packages?

**Comment by ejgallego** [comment]
Other than this curiosity w.r.t. perf numbers, I'm OK with this PR.

**Comment by SkySkimmer** [comment]
>The bench results are pretty surprising, let's see if they are reproducible.

Looks like just noise to me

**Comment by ejgallego** [comment]
> > The bench results are pretty surprising, let's see if they are reproducible.
> 
> Looks like just noise to me

I think the same, but if that's the case our level of noise are worrying, and the use we are doing of the bench too thus.

**Comment by gasche** [comment]
My guess as to what is going on when looking at the number:
- the "CPU cycles" count tell us that nothing has in fact changed
- the "clock time" count looks strange because it suggests that the same benchmarks were sped up across both runs, but my interpretation is that those are benchmarks which had an "unlucky run" in the recorded benchmark results (they did worse in terms of wall-clock than they typically do on this CPU), so running them in loop is likely to (reproductibly) observe improvements

In other words, I guess that the benchmark results that you stored for comparison are just one sample in the runtime distribution of the benchmark programs, so you reproductibly observe improvements if the same was sensibly worse than the mean.

**Comment by gasche** [comment]
I think that you could make the benchmark results more informative by reordering the columns, showing CPU cycles first, then memory, then the more biases "real time" count last.

**Comment by SkySkimmer** [comment]
There are no benchmark results stored for comparison, we run the PR then master for everything we bench.
Master is always changing so I'm not sure how storing results would work.

**Comment by gasche** [comment]
Ah, I assumed that master results were cached. (This would make sense in a model where PR benchmarks run more often than master changes.) Sorry for the noise. Then I don't know why `coq-neural-net-interp-computed-lite` shows up twice as being improved by the chance.

**Comment by ejgallego** [comment]
We should store all the bench results, indeed! 

Maybe let's create an issue so we don't forget to do that.

(Actually I'm not sure how bench stuff is handled these days)

@SkySkimmer something I'm worried about this PR, is that now by default, users pressing Ctrl-C with `coqc` get a bit of uglier result:

```
$ dune exec -- coqc theories/Reals/Rtopology.v 
^CFile "./theories/Reals/Rtopology.v", line 206, characters 0-4:
Error: User interrupt.
```
vs
```
$ dune exec -- coqc theories/Reals/Rtopology.v 
^C
```
Should we tweak the printing when no backtrace?

**Comment by ejgallego** [comment]
@coqbot run full CI

**Comment by ejgallego** [comment]
IDK either, just pointing it out. I wrote uglier due to the `^C` being printed before the line error message, but let me take the ugly back.

I'm always worried about this stuff as who knows what we can break.

@JasonGross , will this impact the bug minimizer?

**Review by ejgallego — APPROVED** [review]

**Comment by JasonGross** [comment]
> Then I don't know why `coq-neural-net-interp-computed-lite` shows up twice as being improved by the chance.

Do we really think this is chance?  The exact same line shows up both times as being improved by 4--8 seconds, and the line is effectively just an invocation of `native_compute` (which I guess on the bench is `vm_compute`).  A long-running call to vm_compute seems like it might be exactly the sort of thing where the signal handler that gets checked matters, if the handler has code that runs frequently?

**Comment by SkySkimmer** [comment]
Does the signal handler matter when there's no signal to handle?

**Comment by ejgallego** [comment]
> > @JasonGross , will this impact the bug minimizer?
> 
> The bug minimizer [terminates the process if it takes too long](https://github.com/JasonGross/coq-tools/blob/8dc4f6ccdd7a96742bc83aa33afc500b6ebb124a/coq_tools/diagnose_error.py#L178-L183), as a way to avoid introducing loops. Additionally the minimizer driver uses `timeout` to avoid hitting GitHub Actions limits. Other than this it does not use signals to communicate with coqc/coqtop, so as long as killing the process is not broken and `timeout N coqc` works, there should not be any issue.

As long as you don't parse the output of coqc _after_ the signal, it should then not matter.

**Comment by ejgallego** [comment]
> > Then I don't know why `coq-neural-net-interp-computed-lite` shows up twice as being improved by the chance.
> 
> Do we really think this is chance? The exact same line shows up both times as being improved by 4--8 seconds, and the line is effectively just an invocation of `native_compute` (which I guess on the bench is `vm_compute`). A long-running call to vm_compute seems like it might be exactly the sort of thing where the signal handler that gets checked matters, if the handler has code that runs frequently?

That's pretty interesting. Indeed the case where we have some external processes could be impacted, but I don't think that's likely as it would imply a really set of weird circumstances.

**Comment by SkySkimmer** [comment]
If terminate is SIGKILL this PR changes nothing for the minimizer

**Comment by SkySkimmer** [comment]
catch_break is SIGINT

**Comment by ejgallego** [comment]
It seems fine to me.

**Comment by ejgallego** [comment]
@coqbot: merge now
